### PR TITLE
Added endpoints for aws sign in url

### PR DESCRIPTION
**Pull request by vigilantee**
Changes for cognito sign in URL generation.

### CODE REVIEW

File: frappe/integrations/oauth2.py [hunk 1]
```diff
@@ -53,7 +53,10 @@ def authorize(*args, **kwargs):
 	params = get_urlparams_from_kwargs(kwargs)
 	request_url = urlparse(frappe.request.url)
 	success_url = request_url.scheme + "://" + request_url.netloc + "/api/method/frappe.integrations.oauth2.approve?" + params
-	failure_url = frappe.form_dict["redirect_uri"] + "?error=access_denied"
+	if("redirect_uri" in frappe.form_dict):
+		failure_url = frappe.form_dict["redirect_uri"] + "?error=access_denied"
+	if("cmd" in frappe.form_dict):
+			failure_url = frappe.form_dict["cmd"] + "?error=access_denied"
```
Comment (hrwX) [inline frappe/integrations/oauth2.py]: - why overriding `failure_url` if the request has `cmd` ?
- `cmd` parameter is mainly consists of dot path to python function

File: frappe/integrations/oauth2_logins.py [hunk 2]
```diff
@@ -35,6 +35,11 @@ def login_via_salesforce(code, state):
 def login_via_fairlogin(code, state):
 	login_via_oauth2("fairlogin", code, state, decoder=decoder_compat)
 
+@frappe.whitelist(allow_guest=True)
+def get_login_url(redirect_to=None, social_login_key_name="bloomstack"):
+	redirect_to = redirect_to.replace("&amp;", "&")
+	return get_oauth2_authorize_url(social_login_key_name, redirect_to)
+
```
Comment (hrwX) [inline frappe/integrations/oauth2_logins.py:42]: - Why not use `custom` function defined for Social Login ?? 
- We can create custom social login in every instance via `bloomtrace_core`.

File: frappe/utils/oauth.py [hunk 3]
```diff
@@ -231,8 +231,8 @@ def update_oauth_user(user, data, provider):
 		user = frappe.new_doc("User")
 		user.update({
 			"doctype":"User",
-			"first_name": get_first_name(data),
-			"last_name": get_last_name(data),
+			"first_name": get_first_name(data) or data["email"],
+			"last_name": get_last_name(data) or "",
```
Comment (hrwX) [inline frappe/utils/oauth.py]:
```suggestion
			"last_name": get_last_name(data),
```
- `last_name` is not mandatory

File: frappe/utils/oauth.py [hunk 4]
```diff
@@ -231,8 +231,8 @@ def update_oauth_user(user, data, provider):
 		user = frappe.new_doc("User")
 		user.update({
 			"doctype":"User",
-			"first_name": get_first_name(data),
-			"last_name": get_last_name(data),
+			"first_name": get_first_name(data) or data["email"],
```
Comment (hrwX) [inline frappe/utils/oauth.py]:
```suggestion
			"first_name": get_first_name(data) or get_email(data),
```

**Review by hrwX — CHANGES_REQUESTED** [review]

File: frappe/integrations/oauth2.py (hunk 1 above)
Comment (vigilantee) [reply frappe/integrations/oauth2.py]: For our use case the redirect_uri is not passed to tackle that we have made this change.

File: frappe/integrations/oauth2_logins.py (hunk 2 above)
Comment (vigilantee) [reply frappe/integrations/oauth2_logins.py:42]: For the cognito sign in we need to get the link for aws sign in and redirect_to is passed in the urlencoded way so we had to intercept the request and make some changes in this.

File: frappe/integrations/oauth2.py (hunk 1 above)
Comment (hrwX) [inline frappe/integrations/oauth2.py]:
```suggestion
	failure_url = (frappe.form_dict["redirect_uri"] or frappe.form_dict["cmd"] or "") + "?error=access_denied"
```

**Review by hrwX — CHANGES_REQUESTED** [review]

**Review by hrwX — APPROVED** [review]

**Review by jitendraharpalani — APPROVED** [review]

**Pull request merged**